File: 2023/haskell/day2.hs
import Data.Char

part1 :: IO ()
part1 = do
    contents <- readFile "input/day2.txt"
    let lines = wordsWhen (=='\n') contents
    print . sumPossibleIds $ map getGame lines

part2 :: IO ()
part2 = do
    contents <- readFile "input/day2.txt"
    let lines = wordsWhen (=='\n') contents
    print . sumPowers $ map getGame lines

sumPowers :: [(Int, [[(Int, String)]])] -> Int
sumPowers [] = 0
sumPowers ((_, sets):xs) = getPower sets + sumPowers xs

getPower :: [[(Int, String)]] -> Int
getPower xs = do
    let (red, green, blue) = getFewestForSet xs
    red * green * blue

getFewestForSet :: [[(Int, String)]] -> (Int, Int, Int)
getFewestForSet [] = (0,0,0)
getFewestForSet (x:xs) = do
    let (red, blue, green) = getFewest x
    let (red', blue', green') = getFewestForSet xs
    (max red red', max blue blue', max green green')

getFewest :: [(Int, String)] -> (Int, Int, Int)
getFewest [] = (0,0,0)
getFewest ((count, color):xs) = do
    let (red, blue, green) = getNeeded count color
    let (red', blue', green') = getFewest xs
    (max red red', max blue blue', max green green')

getNeeded :: Int -> String -> (Int, Int, Int)
getNeeded count color
    | color == "red" = (count, 0, 0)
    | color == "green" = (0, count, 0)
    | otherwise = (0, 0, count)


sumPossibleIds :: [(Int, [[(Int, String)]])] -> Int
sumPossibleIds [(id, sets)]
    | isPossible sets = id
    | otherwise = 0
sumPossibleIds ((id, sets):xs)
    | isPossible sets = id + sumPossibleIds xs
    | otherwise = sumPossibleIds xs

isPossible :: [[(Int, String)]] -> Bool
isPossible [set] = checkSet set
isPossible (set:sets) = checkSet set && isPossible sets

checkSet :: [(Int, String)] -> Bool
checkSet [(count, color)] = checkLimits count color
checkSet ((count, color):xs) = checkLimits count color && checkSet xs

checkLimits :: Int -> String -> Bool
checkLimits count color
    | color == "red" = count <= 12
    | color == "green" = count <= 13
    | otherwise = count <= 14

getGame :: String -> (Int, [[(Int, String)]])
getGame xs = do
    let game = wordsWhen (== ':') xs
    let id = read . getNum $ head game
    let sets = wordsWhen (== ';') (last game)
    let cubes = map getCubes sets
    (id, cubes)

getCubes :: String -> [(Int, String)]
getCubes xs = formatCubes (wordsWhen (==' ') (filter (/=',') xs))

formatCubes :: [String] -> [(Int, String)]
formatCubes [x,y] = [(read x, y)]
formatCubes (x:y:xs) = (read x, y) : formatCubes xs

wordsWhen :: (Char -> Bool) -> String -> [String]
wordsWhen p s =  case dropWhile p s of
    "" -> []
    s' -> w : wordsWhen p s''
        where (w, s'') = break p s'

getNum :: String -> String
getNum [] = []
getNum (x:xs)
    | isDigit x = x:getNum xs
    | otherwise = getNum xs

File: 2023/python/src/day2.py
import os
import sys

def main():
    input = []
    
    with open(f'{os.path.dirname(os.path.realpath(__file__))}/../input/day2.txt') as f:
        for line in f.readlines():
            l = line.replace("\n","")
            input.append(l)

    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))


def part1(input):
    sum = 0
    for i in range(0, len(input)):
        game, sets = input[i].split(":")
        sets = sets.split(";")
        impossible = False
        
        for set in sets:
            bags = set.split(",")
            for bag in bags:
                num, color = filter(len, bag.split(" "))
                if color == "blue":
                    if int(num) > 14:
                        impossible = True
                        break
                if color == "red":
                    if int(num) > 12:
                        impossible = True
                        break 
                if color == "green":
                    if int(num) > 13:
                        impossible = True
                        break 
            if impossible:
                break
            
        if not impossible:
            sum += int(game.split(" ")[1])
    return sum
                

def part2(input):
    sum = 0
    for i in range(0, len(input)):
        game, sets = input[i].split(":")
        sets = sets.split(";")
        
        red = 0
        blue = 0
        green = 0
        
        for set in sets:
            bags = set.split(",")
            for bag in bags:
                num, color = filter(len, bag.split(" "))
                if color == "blue":
                    if int(num) > blue:
                        blue = int(num)
                if color == "red":
                    if int(num) > red:
                        red = int(num)
                if color == "green":
                    if int(num) > green:
                        green = int(num)
            
        sum += (green*red*blue)
        
    return sum



if __name__ == "__main__":
    sys.exit(main())